### ISSUE TITLE
RPC seems to hang indefinitely

### ISSUE THREAD

**Issue by scirner22** [grpc/grpc-java#8334]
Hey! I'm hoping you might have some insights into what could be going on here, or any tips for debugging it better. We have a grpc client on version `1.36.2` and it seems like after some amount of time, threads that issued RPCs hang indefinitely. The calls are being issued with a blocking stub, and in this particular instance they are being wrapped in kotlin coroutines. Here's two traces:

```
"DefaultDispatcher-worker-2" #2112 daemon prio=5 os_prio=0 cpu=77.71ms elapsed=2721.52s tid=0x00007f292c027000 nid=0xd2 waiting on condition  [0x00007f28870f0000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@11.0.11/Native Method)
        - parking to wait for  <0x000000074e238a18> (a java.util.concurrent.CompletableFuture$Signaller)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.11/LockSupport.java:194)
        at java.util.concurrent.CompletableFuture$Signaller.block(java.base@11.0.11/CompletableFuture.java:1796)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@11.0.11/ForkJoinPool.java:3128)
        at java.util.concurrent.CompletableFuture.waitingGet(java.base@11.0.11/CompletableFuture.java:1823)
        at java.util.concurrent.CompletableFuture.get(java.base@11.0.11/CompletableFuture.java:1998)
        at io.p8e.proxy.PermissionUpdaterKt.threadedMap(PermissionUpdater.kt:62)
        at io.p8e.proxy.PermissionUpdater.saveConstructorArguments(PermissionUpdater.kt:24)
        at io.p8e.proxy.Contract.packageContract(Contract.kt:460)
        at io.p8e.proxy.Contract.execute(Contract.kt:719)
        at com.figure.p8e.servicing.service.RetryingP8eContractManager$execute$3.invoke(P8e.kt:162)
        at com.figure.p8e.servicing.service.RetryingP8eContractManager$execute$3.invoke(P8e.kt:161)
        at com.figure.retry.RetryKt.runTaskWithRetryPolicy(Retry.kt:104)
        at com.figure.retry.RetryKt.retrying(Retry.kt:132)
        at com.figure.p8e.servicing.service.RetryingP8eContractManager.execute$suspendImpl(P8e.kt:157)
        at com.figure.p8e.servicing.service.RetryingP8eContractManager.execute(P8e.kt)
        at com.figure.p8e.servicing.fsm.contractstate.AssignLoanToServicingContractFSM.advanceContract(AssignLoanToServicingContract.kt:376)
        at com.figure.p8e.servicing.fsm.contractstate.AssignLoanToServicingContractFSM.access$advanceContract(AssignLoanToServicingContract.kt:54)
        at com.figure.p8e.servicing.fsm.contractstate.AssignLoanToServicingContractFSM$FirstStepAdvanced$nextState$$inlined$invoke$1.invokeSuspend(Effect.kt:49)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)



"DefaultDispatcher-worker-3" #2111 daemon prio=5 os_prio=0 cpu=227.25ms elapsed=2721.52s tid=0x00007f2998028800 nid=0xd1 waiting on condition  [0x00007f28871f1000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@11.0.11/Native Method)
        - parking to wait for  <0x0000000748280000> (a io.grpc.stub.ClientCalls$ThreadlessExecutor)
        at java.util.concurrent.locks.LockSupport.park(java.base@11.0.11/LockSupport.java:194)
        at io.grpc.stub.ClientCalls$ThreadlessExecutor.waitAndDrain(ClientCalls.java:731)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:149)
        at io.p8e.proto.EnvelopeServiceGrpc$EnvelopeServiceBlockingStub.getByExecutionUuid(EnvelopeServiceGrpc.java:488)
        at io.p8e.grpc.client.EnvelopeClient.getByExecutionUuid(EnvelopeClient.kt:37)
        at io.p8e.client.EventMonitorClient.getContract(EngineClient.kt:126)
        at io.p8e.ContractManager.loadContract(ContractManager.kt:348)
        at io.p8e.ContractManager.loadContract(ContractManager.kt:367)
        at com.figure.p8e.servicing.service.RetryingP8eContractManager$loadContract$3.invoke(P8e.kt:172)
        at com.figure.p8e.servicing.service.RetryingP8eContractManager$loadContract$3.invoke(P8e.kt:171)
        at com.figure.retry.RetryKt.runTaskWithRetryPolicy(Retry.kt:104)
        at com.figure.retry.RetryKt.retrying(Retry.kt:132)
        at com.figure.p8e.servicing.service.RetryingP8eContractManager.loadContract-zdnXTc8$suspendImpl(P8e.kt:167)
        at com.figure.p8e.servicing.service.RetryingP8eContractManager.loadContract-zdnXTc8(P8e.kt)
        at com.figure.p8e.servicing.fsm.contractstate.AssignLoanToServicingContractFSM$FirstStepAdvanced$nextState$$inlined$invoke$1.invokeSuspend(Effect.kt:41)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
        at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:750)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
        at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
```

Somewhat ironically, they both have the same elapsed time which seems to mean something happened to cause the whole client to stop processing RPCs. The above are colocated in the logs with GOAWAYs like such

```
Either.Left(PreExecutionError(t=io.grpc.StatusRuntimeException: UNAVAILABLE: Abrupt GOAWAY closed sent stream. HTTP/2 error code:
```

Some added background notes as well:
- A deadline of 60s is on all of these RPCs which doesn't help in this particular instance.
- Datadog APM is attached so they are dynamically injecting some interceptors.
- We have a jwt client interceptor which will make a blocking RPC call for authentication. I've always thought this wasn't the best thing to do, but I've tried to inject faults with it and can't reproduce this blocking behavior.

```
class ChallengeResponseInterceptor(
    private val keyPair: KeyPair,
    private val authenticationClient: AuthenticationClient,
    private val toleranceSeconds: Long = 3
): ClientInterceptor {
    private val jwt = AtomicReference("")

    override fun <ReqT : Any, RespT : Any> interceptCall(
        method: MethodDescriptor<ReqT, RespT>,
        callOptions: CallOptions,
        next: Channel
    ): ClientCall<ReqT, RespT> {
        return object: ForwardingClientCall.SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
            override fun start(
                responseListener: Listener<RespT>,
                headers: Metadata
            ) {
                if (Constant.JWT_CTX_KEY.get() == null) {
                    headers.put(
                        Constant.JWT_METADATA_KEY,
                        jwt.get()
                            .takeIf { it.isNotEmpty() && !it.isExpired() }
                        ?: authenticate())
                }
                super.start(
                    responseListener,
                    headers
                )
            }
        }
    }

    private fun authenticate(): String {
        // blocking RPC call
    }

    private fun String.isExpired(): Boolean {
        return JWT.decode(this)
            .expiresAt
            .toInstant()
            .atOffset(ZoneOffset.UTC)
            .minusSeconds(toleranceSeconds)    // Give some tolerance
            .isBefore(OffsetDateTime.now())
    }
}
```

Any thoughts or debugging help would be much appreciated!

**Comment by dapengzhang0** [comment]
Is `fun authenticate()` hanging? If so, can you please provide more details on `fun authenticate()`?

**Comment by scirner22** [comment]
The authenticate is an rpc for a jwt flow, but the token is only updated every hour. I don't believe that is in play for the traces above, as it shows the rpc being made as this in the second trace.

**Comment by scirner22** [comment]
I can include for completeness though.

```
    private fun authenticate(): String {
        // Random String
        val randomStr = UUID.randomUUID().toString() + System.currentTimeMillis()

        // Generate a token that is value for 10 seconds
        val token = Authentication.AuthenticationToken.newBuilder()
            .setRandomData(randomStr.toByteString())
            .setExpiration(OffsetDateTime.now().plusSeconds(10).toProtoTimestampProv())
            .build()

        // Sign the random token with our private key
        val signature = Signature.getInstance(Constant.JWT_ALGORITHM).apply {
            initSign(keyPair.private)
            update(token.toByteArray())
        }.let {
            it.sign()
        }

        // Authenticate to the server using the 10 second token.
        val jwtResponse = authenticationClient.authenticate(
            Authentication.AuthenticationRequest.newBuilder()
                .setPublicKey(keyPair.public.toPublicKeyProto())
                .setSignature(signature.toByteString())
                .setToken(token)
                .build()
        )

        return jwtResponse.token.also(jwt::set)
    }

```

```
class AuthenticationClient(
    channel: ManagedChannel,
    private val deadlineMs: Long
) {
    private val blockingStub = AuthenticationServiceGrpc
        .newBlockingStub(channel)

    fun authenticate(
        request: Authentication.AuthenticationRequest
    ): Jwt {
        return blockingStub.withDeadlineAfter(deadlineMs, TimeUnit.MILLISECONDS)
            .authenticate(request)
    }
}
```

**Comment by dapengzhang0** [comment]
Can you please add logs to show where it hangs?

For example:
```
                // log here

                if (Constant.JWT_CTX_KEY.get() == null) {
                    headers.put(
                        Constant.JWT_METADATA_KEY,
                        jwt.get()
                            .takeIf { it.isNotEmpty() && !it.isExpired() }
                        ?: authenticate())
                }
                
                // log here

                super.start(
                    responseListener,
                    headers
                )

                // log here
```

**Comment by scirner22** [comment]
I don't have any logs for this because it's in a test system and only happens on rare occasion. Because of how rare it happens, I can't easily reproduce it either. Do you believe this is a problem is the interceptor though? Are there any cases where a rpc call to a blocking stub that's using `withDeadlineAfter(..)` is capable of still blocking after the deadline is exceeded?

**Comment by YifeiZhuang** [comment]
> The authenticate is an rpc for a jwt flow, but the token is only updated every hour. I don't believe that is in play for the traces above, as it shows the rpc being made as this in the second trace.
> 
> ```java
>         at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:149)
>         at io.p8e.proto.EnvelopeServiceGrpc$EnvelopeServiceBlockingStub.getByExecutionUuid(EnvelopeServiceGrpc.java:488)
>         at io.p8e.grpc.client.EnvelopeClient.getByExecutionUuid(EnvelopeClient.kt:37)
> ```

@dapengzhang0 Looking at the log trace detail `.waitAndDrain `,  it seems not the interceptor problem.  Because the `newCall()` is already successfully returned, meaning interceptors are already visited.

https://github.com/grpc/grpc-java/blob/eb6764841b1771e64d222430de7a12c9a6b4ead7/stub/src/main/java/io/grpc/stub/ClientCalls.java#L149

They mentioned that adding a deadline does not help. The deadline executor is using netty eventlookgroup. Also the `GOAWAY` is supposed to close the stream, so it looks the netty thread is hanging somewhere?

**Comment by scirner22** [comment]
The netty thread hanging is what I was originally thinking as well. The other thought I had, and this is more so because I'm not sure how the async model in kotlin coroutines work, is that there were threads never yielding for netty threads to execute.

Do you recommend any suggestions for debugging at the netty level? Is it worth first trying grpc shaded `1.39.0`?

**Comment by dapengzhang0** [comment]
@YifeiZhuang 

> Looking at the log trace detail .waitAndDrain , it seems not the interceptor problem. Because the newCall() is already successfully returned, meaning interceptors are already visited.

.waitAndDrain only indicates `interceptCall()` has returned. But the question is: has the returned ClientCall object finished the `start()` method? That's what I want to make sure.

**Comment by scirner22** [comment]
> @YifeiZhuang
> 
> > Looking at the log trace detail .waitAndDrain , it seems not the interceptor problem. Because the newCall() is already successfully returned, meaning interceptors are already visited.
> 
> .waitAndDrain only indicates `interceptCall()` has returned. But the question is: has the returned ClientCall object finished the `start()` method? That's what I want to make sure.

Is that because there's special logic wrapping the interceptor? If it is the rpc itself or the rpc inside the interceptor is what is hanging, is there a difference? Both of them have 60s deadlines set.

**Comment by YifeiZhuang** [comment]
> > .waitAndDrain only indicates `interceptCall()` has returned. But the question is: has the returned ClientCall object finished the `start()` method? That's what I want to make sure.
> 
> Is that because there's special logic wrapping the interceptor? If it is the rpc itself or the rpc inside the interceptor is what is hanging, is there a difference? Both of them have 60s deadlines set.

@scirner22 
interceptors are run first sequentially, it creates the call. Then the call is started. The RPC deadline takes effect when after call is started. If rpc inside interceptor is hanging, the call won't start at all, there is no connections sent out to the server. If rpc after start is handing, the deadline should take effect unless there is something wrong.

I was saying that the [channel](https://github.com/grpc/grpc-java/blob/master/stub/src/main/java/io/grpc/stub/ClientCalls.java#L142) that the stub is calling is already a [interceptorChannel](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ManagedChannelImpl.java#L680), so, when the `channel.newCall()` returns, the interceptors in your application that blocks the call is already done, so, the rpc is not hanging here but somewhere else.  @dapengzhang0 correct me if I am wrong.

@dapengzhang0  a few lines before .waitAndDrain it shows the created clientCall finishes `start()`.

**Comment by dapengzhang0** [comment]
> a few lines before .waitAndDrain it shows the created clientCall finishes start().

@YifeiZhuang You are right. The start() method has finished. So seems not the problem of the interceptor.

**Comment by scirner22** [comment]
So since the start() method finished, we know that the `waitAndDrain` is waiting for a netty thread to unpark it? A deadline is currently being used so I'm trying to think of what possibly could be going on here? Are there any cases where the netty thread wouldn't service the rpc, or the netty thread itself being starved or something? Attached are the event loop threads from a more recent dump.

```
"grpc-default-worker-ELG-1-1" #43 daemon prio=5 os_prio=0 cpu=11347.60ms elapsed=3251.62s tid=0x00007f3a4e807000 nid=0x34 runnable  [0x00007f39861f9000]
   java.lang.Thread.State: RUNNABLE
        at io.grpc.netty.shaded.io.netty.channel.epoll.Native.epollWait(Native Method)
        at io.grpc.netty.shaded.io.netty.channel.epoll.Native.epollWait(Native.java:148)
        at io.grpc.netty.shaded.io.netty.channel.epoll.Native.epollWait(Native.java:141)
        at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.epollWaitNoTimerChange(EpollEventLoop.java:290)
        at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:347)
        at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.grpc.netty.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.grpc.netty.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(java.base@11.0.12/Thread.java:829)

--
"grpc-default-worker-ELG-1-2" #45 daemon prio=5 os_prio=0 cpu=16868.45ms elapsed=3251.58s tid=0x00007f3a4e673800 nid=0x36 runnable  [0x00007f3985ff7000]
   java.lang.Thread.State: RUNNABLE
        at io.grpc.netty.shaded.io.netty.channel.epoll.Native.epollWait0(Native Method)
        at io.grpc.netty.shaded.io.netty.channel.epoll.Native.epollWait(Native.java:132)
        at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.epollWait(EpollEventLoop.java:286)
        at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:351)
        at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.grpc.netty.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.grpc.netty.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(java.base@11.0.12/Thread.java:829)

"grpc-default-worker-ELG-1-3" #46 daemon prio=5 os_prio=0 cpu=28301.10ms elapsed=3251.09s tid=0x00007f396c08f800 nid=0x37 runnable  [0x00007f3985cf6000]
   java.lang.Thread.State: RUNNABLE
        at io.grpc.netty.shaded.io.netty.channel.epoll.Native.epollWait(Native Method)
        at io.grpc.netty.shaded.io.netty.channel.epoll.Native.epollWait(Native.java:148)
        at io.grpc.netty.shaded.io.netty.channel.epoll.Native.epollWait(Native.java:141)
        at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.epollWaitNoTimerChange(EpollEventLoop.java:290)
        at io.grpc.netty.shaded.io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:347)
        at io.grpc.netty.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:989)
        at io.grpc.netty.shaded.io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)
        at io.grpc.netty.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(java.base@11.0.12/Thread.java:829)

```

**Comment by YifeiZhuang** [comment]
>  So since the start() method finished, we know that the waitAndDrain is waiting for a netty thread to unpark it? 

no, waitAndDrain is blocked on main thread, not the netty thread. It is blocked because there is no response returned to the application from the transport.

> A deadline is currently being used so I'm trying to think of what possibly could be going on here? Are there any cases where the netty thread wouldn't service the rpc, or the netty thread itself being starved or something?

If deadline is successfully installed but not effective, I suspect the thread(netty event group) executing deadline task is hanging.

It looks this becomes duplicate of https://github.com/grpc/grpc-java/issues/8383 regarding MAX_CONCURRENT_STREAM, resolving.
